app/services/medium_providers/default.rb

```diff
@@ -0,0 +1,66 @@
+module MediumProviders
+  class Default < Provider
+    def validate
+      errors = []
+      os = entity.operatingsystem
+      medium = entity.medium
+      arch = entity.architecture
+
+      errors << N_("%{os} medium was not set for host '%{host}'") % { :host => entity, :os => os } if medium.nil?
+      errors << N_("Invalid medium '%{medium}' for '%{os}'") % { :medium => medium, :os => os } unless os.media.include?(medium)
+      errors << N_("Invalid architecture '%{arch}' for '%{os}'") % { :arch => arch, :os => os } unless os.architectures.include?(arch)
+      errors
+    end
+
+    def medium_uri(path = "", &block)
+      url ||= entity.medium.path if entity.medium.present?
+      url ||= ''
+      url += '/' + path unless path.empty?
+      medium_vars_to_uri(url, entity.architecture.name, entity.operatingsystem, &block)
+    end
+
+    def interpolate_vars(pattern)
+      medium_vars_to_uri(pattern, entity.architecture.name, entity.operatingsystem)
+    end
+
+    def unique_id
+      @unique_id ||= begin
+        full_uniq = super
+        "#{entity.medium.name.parameterize}-#{full_uniq[1..10]}"
+      end
+    end
```

Comment:
This could easily be still readable:

```ruby
def unique_id
  @unique_id ||= [entity.try(:medium).try(:name).try(:parameterize), super[1..10]].join(',')
end
```